hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/throttle/ThrottledCollection.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.mongo.throttle
+
+import com.google.inject.{Inject, Singleton}
+import org.mongodb.scala.{MongoClient, Observable, Observer, SingleObservable, Subscription}
+import play.api.{Configuration, Logger}
+
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration.{Duration, DurationInt}
+import scala.util.{Failure, Success}
+
+
+/** Single instance to ensure same throttle is applied across all mongo queries */
+@Singleton
+class ThrottleConfig @Inject()(configuration: Configuration) {
+
+  val maxWaitQueueSize = {
+    val mongoUri = configuration.get[String]("mongodb.uri")
+    val client = MongoClient(uri = mongoUri)
+    client.settings.getConnectionPoolSettings.getMaxWaitQueueSize
+  }
+
+  /** size should be no larger than the WaitQueueSize (default size is 500)
+    * larger than this will cause `com.mongodb.MongoWaitQueueFullException: Too many operations are already waiting for a connection. Max number of operations (maxWaitQueueSize) of 500 has been exceeded.`
+    */
+  val throttleSize =
+    configuration.getOptional[Int]("mongodb.throttlesize").getOrElse(maxWaitQueueSize)
+
+  Logger.debug(s"Throttling mongo queries using throttleSize=$throttleSize")
+
+  val throttledEc =
+    ExecutionContext.fromExecutorService(
+      java.util.concurrent.Executors.newFixedThreadPool(throttleSize))
+
+  val timeout =
+    configuration.getOptional[Duration]("mongodb.queryTimeout").getOrElse(20.seconds)
```

Comment:
Just curious, is 20s the default in mongo, or that's just a good value we've found?